azavea/ashlar-blueprint
# Initialize dev environment with a running Ashlar instance

## Overview

This PR sets up a dev environment for the Ashlar Blueprint, and gets it going by starting a server running an Ashlar instance.

### Demo

Run `docker-compose run --service-ports ashlar-server`, navigate to `localhost:8000/api`, and voilà:

![api-root](https://user-images.githubusercontent.com/14170650/41294228-819988a2-6e25-11e8-9b74-a00a0235fa79.png)

### Notes

Getting `docker-compose` to serve the app properly was a pain, but I learned a lot in the process. Two specific details that took me a while to figure out:

- Although `docker-compose` automatically makes all services discoverable to one another on a local network, "services" are _containers_, not _images_, meaning that image build scripts (Dockerfiles) can't depend on reaching other services over the network. At first I was trying to use the Ashlar instance Dockerfile to run migrations against the database, but it couldn't resolve the `db` hostname because that host gets set up for the running containers, not the images that they're spun up from.
- `docker-compose up` and `docker-compose run` are not equivalent commands. In particular, `docker-compose run` doesn't create port mappings defined in `docker-compose.yml` by default -- you have to specify that you want the ports by adding the `--service-ports` flag.
 
In spite of this, there are a few points of `docker-compose` that I'm still confused about. I got them to work, but the code is a little superstitious since I don't fully understand what's going on. Those points include:

- For some reason, I needed to explicitly specify that I wanted to map ports of the loopback interface with the `ports` directive (i.e. mapping `ports: - "8000"` or `ports: - "8000:8000"` didn't expose the app on `localhost:8000` on the host, but mapping `ports: - "127.0.0.1:8000:8000"` did). I still don't quite understand why this is, since [the docs for `ports` are pretty sparse](https://docs.docker.com/compose/compose-file/#ports). 
- Even though I specified the `db` service as a dependency for the `ashlar-server` service (twice, since the `links` and `depends_on` directives both influence the order of setup), there are a few occasions in which the database server doesn't get set up in time for `ashlar-server` to reach it when it needs to. Specifically, running `docker-compose up` (as opposed to `docker-compose run`) seems to always spin the two services up in parallel, which causes `ashlar-server` to complain about not being able to access the database; another issue is that when running `docker-compose run` when neither service is started, `docker-compose` will sometimes start them simultaneously. The one situation that always works is when the `db` service has already been started, in which case `docker-compose run` works fine.
 
## Testing Instructions

Follow the instructions in the root README for making sure you have the right Docker requirements. Then, build and run the services:

```
docker-compose run --service-ports ashlar-server
```

You should be able to navigate to `localhost:8000/api` and view a running Ashlar instance.
